src/libos/src/process/do_vfork.rs

```diff
@@ -0,0 +1,168 @@
+// to new child process execve or exit, the child process just reuse the parent process's everything, including
+// task, pid and etc. And also the log of child process will not start from the point that vfork returns but the
+// point that execve returns.
+
+lazy_static! {
+    // Track parent and his vforked child. A parent only has one vforked child at a time.
+    // K: parent pid, V: child pid
+    static ref VFORK_PROCESS_TABLE: SgxMutex<HashMap<pid_t, pid_t>> = SgxMutex::new(HashMap::new());
+    // Store all the parents's file tables who call vfork. It will be recovered when the child exits or has its own task.
+    // K: parent pid, V: parent file table
+    static ref VFORK_PARENT_FILE_TABLES: SgxMutex<HashMap<pid_t, FileTable>> = SgxMutex::new(HashMap::new());
+}
+
+thread_local! {
+    // Store the current thread's cpu context in thread local storage when current thread calls vfork.
+    static VFORK_CPUCONTEXT: RefCell<Option<CpuContext>> = Default::default();
+}
+
+pub fn do_vfork(mut context: *mut CpuContext) -> Result<isize> {
+    let current = current!();
+    trace!("vfork parent process pid = {:?}", current.process().pid());
+
+    // Generate a new pid for child process
+    let new_pid = {
+        let new_tid = ThreadId::new();
+        new_tid.as_u32() as pid_t
+    };
+
+    // Store parent and the vforked child
+    let parent_pid = current.process().pid();
+    let mut vfork_process_table = VFORK_PROCESS_TABLE.lock().unwrap();
+    if let Some(_) = vfork_process_table.insert(parent_pid, new_pid) {
+        return_errno!(EINVAL, "current process's vfork has not returned yet");
+    }
+
+    // Save parent's user context in TLS
+    VFORK_CPUCONTEXT.with(|cell| {
+        let mut p_context = cell.borrow_mut();
+        *p_context = unsafe { Some((*context).clone()) };
+    });
+
+    // Save parent's file table
+    let mut vfork_file_tables = VFORK_PARENT_FILE_TABLES.lock().unwrap();
+    let parent_file_table = {
+        let mut current_file_table = current.files().lock().unwrap();
+        let new_file_table = current_file_table.clone();
+        // FileTable contains non-cloned struct, so here we do a memory replacement.
+        mem::replace(&mut *current_file_table, new_file_table)
+    };
+    if let Some(_) = vfork_file_tables.insert(parent_pid, parent_file_table) {
+        return_errno!(EINVAL, "current process's vfork has not returned yet");
+    }
+
+    // This is the first time return and will return as child.
+    // The second time return will return as parent in vfork_return_to_parent.
+    info!("vfork child pid = {:?}", new_pid);
+    return Ok(0 as isize);
+}
+
+// Check if the calling process is a vforked child process that reuse parent's task and pid.
+pub fn is_vforked_child_process() -> bool {
+    // Due to current limitation, the child process is reusing parent process's task and pid.
+    // Thus parent pid is current pid.
+    let parent_pid = current!().process().pid();
```

Comment:
Can we use thread local storage to make it simple?

Reply:
Store child_pid in TLS but filetable can't be put to TLS because filetable has struct with Arc and is contradictory with RefCell.